src/createEventEmitterProxy.ts

```diff
@@ -0,0 +1,103 @@
+ * Creates a proxy object that initially points to the given object but whose
+ * target can be substituted with another object later using its `setTarget`
+ * method. In addition, when the target is changed, event listeners which have
+ * been attached to the target will be detached and migrated to the new target.
+ *
+ * @param initialTarget - The initial object you want to wrap.
+ * @param opts - The options.
+ * @param opts.eventFilter - Usually, listeners for all events will be migrated
+ * from old targets to new targets, but this function can be used to select
+ * which events you want to migrate. If you pass `'skipInternal'`, then
+ * `newListener` and `removeListener` will be excluded.
+ * @returns The proxy object.
+ */
+export function createEventEmitterProxy<T extends EventEmitter>(
+  initialTarget: T,
+  {
+    eventFilter: givenEventFilter = filterNoop,
+  }: {
+    eventFilter?: ((eventName: string | symbol) => boolean) | 'skipInternal';
+  } = {},
+): SwappableProxy<T> {
+  // parse options
+  const eventFilter =
+    givenEventFilter === 'skipInternal'
+      ? externalEventFilter
+      : givenEventFilter;
+
+  if (typeof eventFilter !== 'function') {
+    throw new Error('createEventEmitterProxy - Invalid eventFilter');
+  }
+
+  let target = initialTarget;
+
+  /**
+   * Changes the object that the proxy wraps.
+   *
+   * @param newTarget - The new object.
+   */
+  let setTarget = (newTarget: any) => {
+    const oldTarget = target;
+    target = newTarget;
+    // migrate listeners
+    oldTarget
+      .eventNames()
+      .filter(eventFilter)
+      .forEach((name) => {
+        oldTarget
+          .rawListeners(name)
+          .forEach((handler) => newTarget.on(name, handler));
+      });
+    // remove old
+    oldTarget.removeAllListeners();
+  };
+
+  const proxy: SwappableProxy<T> = new Proxy<any>(target, {
+    get: (_, name, receiver: T) => {
+      // override `setTarget` access
+      if (name === 'setTarget') {
+        return setTarget;
+      }
+
+      // Typecast: We cannot typecast `name` in the arguments for this option
+      // because it conflicts with the function signature, so we need to
+      // typecast `target` instead.
+      const value = (target as any)[name];
+      if (value instanceof Function) {
+        return function (this: T | undefined, ...args: any[]) {
+          // This function may be a method bound to the proxy object, or an
+          // unbound function.
+          // eslint-disable-next-line no-invalid-this
+          return value.apply(this === receiver ? target : this, args);
+        };
+      }
+      return value;
+    },
+    set: (_, name, value) => {
+      // allow `setTarget` overrides
+      if (name === 'setTarget') {
+        setTarget = value;
+        return true;
+      }
+      // Typecast: We cannot typecast `name` in the arguments for this option
+      // because it conflicts with the function signature, so we need to
+      // typecast `target` instead.
```

Comment:
Nit: I don't think this comment applies, because typecasting `name` wouldn't work even if it was an option. This line is modifying `target`, so it conflicts with using `T` as the type for `target`. We need to cast it to `any` here either way.

```suggestion
      // Typecast: Cast to `any` to allow arbitrary modifications to `target`
```

Reply:
Actually.... maybe this means we should set the type of `target` to `any` as well. e.g.

```
let target: any = initialTarget;
```

That would be more accurate; it really _can_ be anything. And it would remove another cast, sorta.

Reply:
I've removed this in recent commits.